src/coreclr/jit/lowerxarch.cpp

```diff
@@ -2689,6 +2694,144 @@ GenTree* Lowering::LowerHWIntrinsicCndSel(GenTreeHWIntrinsic* node)
+                assert(ternLogInfo.oper3 == TernaryLogicOperKind::Cond);
+
+                GenTree* condition   = nullptr;
+                GenTree* selectTrue  = nullptr;
+                GenTree* selectFalse = nullptr;
+
+                if (ternLogInfo.oper1Use == TernaryLogicUseFlags::A)
+                {
+                    selectTrue = op1;
+
+                    if (ternLogInfo.oper2Use == TernaryLogicUseFlags::B)
+                    {
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::C);
+
+                        selectFalse = op2;
+                        condition   = op3;
+                    }
+                    else
+                    {
+                        assert(ternLogInfo.oper2Use == TernaryLogicUseFlags::C);
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::B);
+
+                        selectFalse = op3;
+                        condition   = op2;
+                    }
+                }
+                else if (ternLogInfo.oper1Use == TernaryLogicUseFlags::B)
+                {
+                    selectTrue = op2;
+
+                    if (ternLogInfo.oper2Use == TernaryLogicUseFlags::A)
+                    {
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::C);
+
+                        selectFalse = op1;
+                        condition   = op3;
+                    }
+                    else
+                    {
+                        assert(ternLogInfo.oper2Use == TernaryLogicUseFlags::C);
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::A);
+
+                        selectFalse = op3;
+                        condition   = op1;
+                    }
+                }
+                else
+                {
+                    assert(ternLogInfo.oper1Use == TernaryLogicUseFlags::C);
+
+                    selectTrue = op3;
+
+                    if (ternLogInfo.oper2Use == TernaryLogicUseFlags::A)
+                    {
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::B);
+
+                        selectFalse = op1;
+                        condition   = op2;
+                    }
+                    else
+                    {
+                        assert(ternLogInfo.oper2Use == TernaryLogicUseFlags::B);
+                        assert(ternLogInfo.oper3Use == TernaryLogicUseFlags::A);
+
+                        selectFalse = op2;
+                        condition   = op1;
+                    }
+                }
+
+                if (!condition->OperIsHWIntrinsic(NI_AVX512F_ConvertMaskToVector))
+                {
+                    break;
+                }
+
+                node->ResetHWIntrinsicId(NI_AVX512F_BlendVariableMask, comp, selectFalse, selectTrue,
+                                         condition->AsHWIntrinsic()->Op(1));
+
+                BlockRange().Remove(op1);
```

Comment:
is it possible that you're deleting nodes with side-effects?

Reply:
It's explicitly checked to be the hwintrinsic node ConvertVectorToMask, so it shouldn't ever had any side effects

Reply:
Ah, you know what, though. This is deleting the wrong thing and should be removing `condition` not `op1`. Let me fix that